File: backend/app/schemas/ml.py
from typing import Dict, List, Optional, Any, Union
from datetime import datetime
from pydantic import BaseModel, Field


class MLModelBase(BaseModel):
    """Base ML model schema with common attributes."""
    name: str
    algorithm: str
    type: str = Field(..., description="Model type (classification, regression)")
    description: Optional[str] = None
    target_variable: str


class MLModelCreate(MLModelBase):
    """Schema for ML model creation."""
    project_id: Optional[int] = None
    dataset_ids: List[int] = Field(default_factory=list)
    hyperparameters: Optional[Dict[str, Any]] = None


class MLModelUpdate(BaseModel):
    """Schema for ML model updates - all fields optional."""
    name: Optional[str] = None
    description: Optional[str] = None
    status: Optional[str] = None
    dataset_ids: Optional[List[int]] = None
    hyperparameters: Optional[Dict[str, Any]] = None


class MLModelInDB(MLModelBase):
    """ML model schema as stored in database."""
    id: str
    created_at: datetime
    updated_at: datetime
    status: str
    project_id: Optional[int] = None
    hyperparameters: Optional[Dict[str, Any]] = None
    metrics: Optional[Dict[str, float]] = None
    feature_importances: Optional[Dict[str, float]] = None
    model_file_path: Optional[str] = None
    
    class Config:
        orm_mode = True


class MLModel(MLModelInDB):
    """Full ML model schema for API responses."""
    dataset_ids: List[int] = Field(default_factory=list)


class Feature(BaseModel):
    """Schema for model features."""
    name: str
    importance: float
    category: Optional[str] = None


class FeatureImportance(BaseModel):
    """Schema for feature importance."""
    model_id: str
    features: List[Feature]


class AlgorithmParameter(BaseModel):
    """Schema for algorithm parameters."""
    name: str
    type: str
    default: Any
    min: Optional[float] = None
    max: Optional[float] = None
    options: Optional[List[str]] = None
    description: Optional[str] = None


class Algorithm(BaseModel):
    """Schema for ML algorithms."""
    id: str
    name: str
    type: List[str]
    description: str
    parameters: List[AlgorithmParameter]


class ClassificationPrediction(BaseModel):
    """Schema for classification predictions."""
    id: Union[int, str]
    prediction: Union[int, str]
    probability: Optional[float] = None


class RegressionPrediction(BaseModel):
    """Schema for regression predictions."""
    id: Union[int, str]
    prediction: float


class PredictionResult(BaseModel):
    """Schema for prediction results."""
    model_id: str
    predictions: List[Union[ClassificationPrediction, RegressionPrediction]]


class PreparedDataCreate(BaseModel):
    """Schema for prepared data creation."""
    dataset_id: int
    target_column: str
    feature_columns: Optional[List[str]] = None
    network_metrics: Optional[List[str]] = None
    network_id: Optional[int] = None
    test_size: Optional[float] = 0.2


class PreparedData(BaseModel):
    """Schema for prepared data."""
    id: int
    dataset_id: int
    target_column: str
    feature_columns: Optional[List[str]] = None
    network_metrics: Optional[List[str]] = None
    network_id: Optional[int] = None
    test_size: float = 0.2
    created_at: datetime
    file_path: Optional[str] = None
    
    class Config:
        orm_mode = True

File: frontend/src/types/ml.ts
export type ModelType = 'classification' | 'regression';
export type ModelStatus = 'created' | 'trained' | 'failed';
export type AlgorithmType = 'random_forest' | 'logistic_regression' | 'linear_regression' | 'svm' | 'neural_network';

export interface MLModel {
  id: number;
  name: string;
  description?: string;
  created_at: string;
  updated_at: string;
  type: ModelType;
  algorithm: AlgorithmType;
  target_variable: string;
  project_id?: number;
  dataset_ids: number[];
  network_id?: number;
  status: ModelStatus;
  metrics?: Record<string, number>;
}

export interface Feature {
  name: string;
  importance: number;
  category: string;
}

export interface FeatureImportance {
  model_id: number;
  features: Feature[];
}

export interface AlgorithmParameter {
  name: string;
  type: 'int' | 'float' | 'categorical' | 'tuple';
  default: any;
  min?: number;
  max?: number;
  options?: string[];
  description?: string;
}

export interface Algorithm {
  id: AlgorithmType;
  name: string;
  type: ModelType[];
  description: string;
  parameters: AlgorithmParameter[];
}

export interface ClassificationPrediction {
  id: number | string;
  prediction: number | string;
  probability?: number;
}

export interface RegressionPrediction {
  id: number | string;
  prediction: number;
}

export interface PredictionResult {
  model_id: number;
  predictions: Array<ClassificationPrediction | RegressionPrediction>;
}

export interface TrainingOptions {
  train_test_split?: number;
  cross_validation?: {
    method: 'k_fold' | 'stratified_k_fold' | 'leave_one_out' | 'time_series';
    k?: number;
  };
  class_weights?: 'balanced' | 'none' | Record<string | number, number>;
  hyperparameter_tuning?: {
    enabled: boolean;
    method: 'grid_search' | 'random_search';
    parameters?: Record<string, any[]>;
    cv?: number;
  };
  algorithm_parameters?: Record<string, any>;
}